source/posix/rw_lock.c

```diff
@@ -12,48 +12,103 @@
+        aws_semaphore_clean_up(&lock->reader_sem);
+        return AWS_OP_ERR;
+    }
+    if (aws_mutex_init(&lock->writer_lock)) {
+        aws_semaphore_clean_up(&lock->reader_sem);
+        aws_semaphore_clean_up(&lock->writer_sem);
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_ERR;
 }
 
 void aws_rw_lock_clean_up(struct aws_rw_lock *lock) {
 
-    pthread_rwlock_destroy(&lock->lock_handle);
+    aws_semaphore_clean_up(&lock->reader_sem);
+    aws_semaphore_clean_up(&lock->writer_sem);
+    aws_mutex_clean_up(&lock->writer_lock);
 }
 
 int aws_rw_lock_rlock(struct aws_rw_lock *lock) {
 
-    return aws_private_convert_and_raise_error_code(pthread_rwlock_rdlock(&lock->lock_handle));
+    const reader_count_t num_readers = aws_atomic_fetch_add(&lock->readers, 1) + 1;
+    if (num_readers < 0) {
+        aws_semaphore_acquire(&lock->reader_sem);
+    }
+    assert(aws_atomic_load_int(&lock->readers) > 0);
+
+    return AWS_OP_SUCCESS;
 }
 
 int aws_rw_lock_wlock(struct aws_rw_lock *lock) {
 
-    return aws_private_convert_and_raise_error_code(pthread_rwlock_wrlock(&lock->lock_handle));
-}
+    aws_mutex_lock(&lock->writer_lock);
 
-int aws_rw_lock_try_rlock(struct aws_rw_lock *lock) {
+    const reader_count_t num_readers = aws_atomic_fetch_sub(&lock->readers, s_max_readers);
+    if (num_readers > 0) {
+        const reader_count_t num_holdouts = aws_atomic_fetch_add(&lock->holdouts, num_readers) + num_readers;
+        assert(num_holdouts >= 0);
+        if (num_holdouts > 0) {
+            aws_semaphore_acquire(&lock->writer_sem);
+        }
+        assert(aws_atomic_load_int(&lock->holdouts) == 0);
+    }
 
-    return aws_private_convert_and_raise_error_code(pthread_rwlock_tryrdlock(&lock->lock_handle));
+    return AWS_OP_SUCCESS;
 }
 
-int aws_rw_lock_try_wlock(struct aws_rw_lock *lock) {
+int aws_rw_lock_runlock(struct aws_rw_lock *lock) {
 
-    return aws_private_convert_and_raise_error_code(pthread_rwlock_trywrlock(&lock->lock_handle));
-}
+    const reader_count_t num_readers = aws_atomic_fetch_sub(&lock->readers, 1) - 1;
 
-int aws_rw_lock_runlock(struct aws_rw_lock *lock) {
+    if (num_readers < 0) {
+        const reader_count_t num_holdouts = aws_atomic_fetch_sub(&lock->holdouts, 1) - 1;
+        if (num_holdouts == 0) {
+            aws_semaphore_release_one(&lock->writer_sem);
+        }
+    }
 
-    return aws_private_convert_and_raise_error_code(pthread_rwlock_unlock(&lock->lock_handle));
+    return AWS_OP_SUCCESS;
 }
 
 int aws_rw_lock_wunlock(struct aws_rw_lock *lock) {
 
-    return aws_private_convert_and_raise_error_code(pthread_rwlock_unlock(&lock->lock_handle));
+    assert(aws_atomic_load_int(&lock->holdouts) == 0);
+
+    const reader_count_t current = aws_atomic_fetch_add(&lock->readers, s_max_readers) + s_max_readers;
+    assert(current >= 0);
+
+    for (reader_count_t i = 0; i < current; ++i) {
+        aws_semaphore_release_one(&lock->reader_sem);
```

Comment:
Why not introduce a release_multiple API to avoid reacquiring the internal semaphore mutex multiple times?

Reply:
That's a good idea, and shouldn't be too hard.